pandas/tests/test_frame.py

```diff
@@ -12385,6 +12404,93 @@ def verify(df):
                 for col in ['4th', '5th']:
                     verify(udf[col])
 
+        # GH7403
+        df = pd.DataFrame({'A': list('aaaabbbb'),'B':range(8), 'C':range(8)})
+        df.iloc[3, 1] = np.NaN
+        left = df.set_index(['A', 'B']).unstack(0)
+
+        vals = [[3, 0, 1, 2, nan, nan, nan, nan],
+                [nan, nan, nan, nan, 4, 5, 6, 7]]
+        vals = list(map(list, zip(*vals)))
+        idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')
```

Comment:
The question is what you want the behavior to be. Suppose ``df.index`` is a 2-level ``MultiIndex``, the first level of which is (effectively) an ``Int64Index`` with missing values (i.e. ``-1`` in ``labels``), and that you then unstack the second level (i.e. ``df.unstack(1)``). What should the result ``index`` be? There are two options: (a) ``Float64Index``, and (b) ``Index(..., dtype='object')``. With this PR, the result is (a). Personally I think (b) is preferable.

```
In : df = pd.DataFrame([[1,2],[3,4]], index=pd.MultiIndex.from_tuples([(1, 'a'), (None, 'b')], names=['ints', 'letters']))
In : result = df.unstack(1)
In : result.index
Out: Float64Index([nan, 1.0], dtype='float64')
```